README.md

```diff
@@ -1,105 +1,273 @@
+        return len(self.genome)
+```
+**\_\_len\_\_ complexity:** $O(1)$
+___
+```python
+def __str__(self) -> str:
+        return "".join(self.genome)
+```
+**\_\_str\_\_ complexity:** $O(n)$
+___
+
+### Doubly-linked list
+
+```python
+def __init__(self, n: int):
+        self.head = Link(None, None, None) # O(1)
+        self.head.next = self.head # O(1)
+        self.head.prev = self.head # O(1)
+        self.length = n # O(1)
+        self.te_identities = {} # O(1)
+        self.te_count = 0 # O(1)
+        for _ in range(n): # O(n)
+            insert_after(self.head.prev, "-") # O(1)
+```
+**\_\_init\_\_ complexity:** $O(n)$
+___
+```python
+def insert_te(self, pos: int, length: int) -> int:
+        active_tes = list(self.te_identities.items()) # O(k)
+        for id, te_range in active_tes: # O(k)
+            start, end = te_range # O(1)
+            if start <= pos < end: # O(1)
+                self.disable_te(id)
+                continue
+            if start > pos: # O(1)
+                self.te_identities[id] = (start + length, end + length) # O(1)
+        self.te_count += 1 # O(1)
+        self.te_identities[self.te_count] = (pos, pos + length) # O(1)
+        current_link = self.head.next # O(1)
+        for _ in range(pos - 1): # O(n)
+            current_link = current_link.next # O(1)
+        for _ in range(length): # O(m)
+            insert_after(current_link, 'A') # O(1)
+            current_link = current_link.next # O(1)
+        self.length = self.length + length # O(1)
+        return self.te_count
+```
+**insert_te complexity:** $O(k+n+m)$
+___
+```python
+def copy_te(self, te: int, offset: int) -> int | None:
+        if self.te_identities[te]: # O(1)
+            te_range = self.te_identities[te] # O(1)
+            start, end = te_range # O(1)
+            copy_length = end - start # O(1)
+            copy_pos = start + offset # O(1)
+            if offset > 0: # O(1)
+                return self.insert_te(copy_pos, copy_length)
+            if (start + offset) < 0: # O(1)
+                copy_pos = self.length + start + offset # O(1)
+                return self.insert_te(copy_pos, copy_length)
+            copy_pos = start + offset # O(1)
+            return self.insert_te(copy_pos, copy_length)
+```
+**copy_te complexity:** $O(1)/O(k + n + m))$
+
+While most of the copy-function can be completed in constant time, it also inherits the complexity of the insert function
+___
+```python
+def disable_te(self, te: int) -> None:
+        if self.te_identities[te]: # O(1)
+            te_range = self.te_identities.pop(te) # O(1) average / O(n) amortized worst case
+            start, end = te_range # O(1)
+            current_link = self.head.next # O(1)
+            for _ in range(start): # O(n)
+                current_link = current_link.next # O(1)
+            for _ in range(end - start): # O(m)
+                current_link.val = 'x' # O(1)
+                current_link = current_link.next # O(1)

```

Comment:
If you didn't keep the genome string updated throughout, this could be a $O(1)$ operation. If you have a map of active identifiers, then you just have to remove `te` from it. Then, when you print the genome in `__str__`, you just have to map to `A` or `x` depending on whether the ID at a position is disabled or not. Since you cannot do `__str__` faster than $O(n)$, that would not change its complexity.

But this is an okay solution.